lib/ui/painting/picture.cc

```diff
@@ -16,6 +18,54 @@
 #include "third_party/tonic/logging/dart_invoke.h"
 
 namespace flutter {
+namespace {
+
+sk_sp<SkSurface> MakeSnapshotSurface(const SkISize& picture_size,
+                                     fml::WeakPtr<GrContext> resource_context) {
+  SkImageInfo image_info = SkImageInfo::MakeN32Premul(
+      picture_size.width(), picture_size.height(), SkColorSpace::MakeSRGB());
+  if (resource_context) {
+    return SkSurface::MakeRenderTarget(resource_context.get(),  // context
+                                       SkBudgeted::kNo,         // budgeted
+                                       image_info               // image info
+    );
+  } else {
+    return SkSurface::MakeRaster(image_info);
+  }
+}
+
+sk_sp<SkImage> MakeRasterSnapshot(sk_sp<SkPicture> picture,
+                                  sk_sp<SkSurface> surface) {
+  TRACE_EVENT0("flutter", __FUNCTION__);
+
+  if (surface == nullptr || surface->getCanvas() == nullptr) {
+    return nullptr;
+  }
+
+  surface->getCanvas()->drawPicture(picture.get());
+
+  surface->getCanvas()->flush();
+
+  sk_sp<SkImage> device_snapshot;
+  {
+    TRACE_EVENT0("flutter", "MakeDeviceSnpashot");
+    device_snapshot = surface->makeImageSnapshot();
+  }
+
+  if (device_snapshot == nullptr) {
+    return nullptr;
+  }
+
+  {
+    TRACE_EVENT0("flutter", "DeviceHostTransfer");
```

Comment:
Maybe leave a comment here like "Here device means GPU and host means CPU; this is different from use cases like Flutter driver tests where device means mobile devices and host means laptops/desktops."

Reply:
Done.